# origin: chrisw (BitKeeper)
# cset: 1.1982.161.34 (2.6) key=4244bedbODYVU4s1YAGRvZSt28TAgA
# URL: http://linux.bkbits.net:8080/linux-2.6/cset@4244bedbODYVU4s1YAGRvZSt28TAgA
# inclusion: upstream
# descrition: [PATCH] isofs: more defensive checks against corrupt isofs images
# revision date: Mon, 28 Mar 2005 18:49:17 +0900
#
# S rset: ChangeSet|1.1982.161.33..1.1982.161.34
# I rset: fs/isofs/rock.c|1.22.1.1..1.22.1.2
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
# R: Revised  Manually revised by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/03/25 17:46:03-08:00 chrisw@osdl.org 
#   [PATCH] isofs: more defensive checks against corrupt isofs images
#   
#   Michal Zalewski <lcamtuf@dione.ids.pl> discovers range checking flaws in
#   iso9660 filesystem.
#   
#   http://marc.theaimsgroup.com/?l=bugtraq&m=111110067304783&w=2
#   
#   CAN-2005-0815 is assigned to this issue.
#   
#   Some more defensive checks to keep corrupt isofs images from corrupting
#   memory or causing Oops.
#   
#   Signed-off-by: Chris Wright <chrisw@osdl.org>
#   
#   ===== fs/isofs/rock.c 1.23 vs edited =====
# 
# fs/isofs/rock.c
#   2005/03/24 00:41:31-08:00 chrisw@osdl.org +4 -0
#   isofs: more defensive checks against corrupt isofs images
# 
#
===== fs/isofs/rock.c 1.22.1.1 vs 1.22.1.2 =====
--- 1.22.1.1/fs/isofs/rock.c	2005-03-26 08:25:50 +09:00
+++ 1.22.1.2/fs/isofs/rock.c	2005-03-24 17:41:31 +09:00
@@ -74,6 +74,10 @@
     offset1 = 0; \
     pbh = sb_bread(DEV->i_sb, block); \
     if(pbh){       \
+      if (offset > pbh->b_size || offset + cont_size > pbh->b_size){	\
+	brelse(pbh); \
+	goto out; \
+      } \
       memcpy(buffer + offset1, pbh->b_data + offset, cont_size - offset1); \
       brelse(pbh); \
       chr = (unsigned char *) buffer; \
